### PR TITLE
Track source more generically in ReactComponentTreeDevtool

### PR DESCRIPTION
**Pull request by spicyj**
Being able to get the source for your parent components seems useful, and ReactComponentTreeDevtool is best poised to be able to do that.

I'm also not sure it makes sense to have separate DOM-specific `onMountDOMComponent` and `onUpdateDOMComponent` events, so I removed them for now. Even if we want them, their timing seemed sort of arbitrary.

I also made it so DOM devtools can listen to non-DOM events too. Willing to change that if people think it's ugly though.

Reviewers: @gaearon

### CODE REVIEW

File: src/renderers/dom/shared/ReactDOMDebugTool.js [hunk 1]
```diff
@@ -40,9 +41,11 @@ function emitEvent(handlerFunctionName, arg1, arg2, arg3, arg4, arg5) {
 
 var ReactDOMDebugTool = {
   addDevtool(devtool) {
+    ReactDebugTool.addDevtool(devtool);
```
Comment (gaearon) [inline src/renderers/dom/shared/ReactDOMDebugTool.js]: Are we going to expose just one of them in the end?

File: src/renderers/dom/shared/devtools/ReactDOMUnknownPropertyDevtool.js [hunk 2]
```diff
@@ -47,14 +48,18 @@ if (__DEV__) {
         null
     );
 
+    var source = ReactComponentTreeDevtool.getSource(lastDebugID);
+    var formattedSource = source ?
```
Comment (gaearon) [inline src/renderers/dom/shared/devtools/ReactDOMUnknownPropertyDevtool.js]: Is it worth adding `getFormattedSource()` to `ReactComponentTreeDevtool` for convenience?

File: src/isomorphic/devtools/ReactComponentTreeDevtool.js [hunk 3]
```diff
@@ -19,6 +19,7 @@ var rootIDs = [];
 function updateTree(id, update) {
   if (!tree[id]) {
     tree[id] = {
+      source: null,
```
Comment (gaearon) [inline src/isomorphic/devtools/ReactComponentTreeDevtool.js]: While we’re at it, we might as well store the `element` itself because React DevTools will want to know it.

File: src/isomorphic/ReactDebugTool.js [hunk 4]
```diff
@@ -31,7 +31,7 @@ function emitEvent(handlerFunctionName, arg1, arg2, arg3, arg4, arg5) {
           handlerDoesThrowForEvent[handlerFunctionName],
           'exception thrown by devtool while handling %s: %s',
           handlerFunctionName,
-          e.message
+          e.message + '\n' + e.stack
```
Comment (gaearon) [inline src/isomorphic/ReactDebugTool.js]: We should do this in `ReactDOMDebugTool` for symmetry. You can extract it if you’d like.

Comment (spicyj) [inline src/isomorphic/ReactDebugTool.js]: Oops, this was just for debugging. We can keep it though; it seems useful.

Comment (gaearon) [inline src/isomorphic/ReactDebugTool.js]: Yeah, I agree it’s useful. Would you like to do it as a separate PR?

File: src/renderers/dom/shared/ReactDOMDebugTool.js (hunk 1 above)
Comment (spicyj) [inline src/renderers/dom/shared/ReactDOMDebugTool.js]: I figured we would probably expose both – though maybe it makes more sense to only expose each platform-specific one. Open to other suggestions on how to set ReactDOMUnknownPropertyDevtool up so that we can listen to both kinds of events though.

File: src/isomorphic/devtools/ReactComponentTreeDevtool.js (hunk 3 above)
Comment (spicyj) [inline src/isomorphic/devtools/ReactComponentTreeDevtool.js]: I think storing element seems reasonable enough, but I believe React Devtools is set up to store its own tree so I'm not actually sure this would help it. Maybe I'm misremembering.

File: src/renderers/dom/shared/ReactDOMDebugTool.js (hunk 1 above)
Comment (gaearon) [inline src/renderers/dom/shared/ReactDOMDebugTool.js]: >maybe it makes more sense to only expose each platform-specific one

Let’s go with this assumption for now. In this case it makes sense to automatically setup forwarding.

File: src/isomorphic/devtools/ReactComponentTreeDevtool.js (hunk 3 above)
Comment (gaearon) [inline src/isomorphic/devtools/ReactComponentTreeDevtool.js]: Yes, but AFAIK we plan to kill that code and make React DevTools use this tool instead. (Or implement a similar one using the same events.) For older versions of React, we will add a shim to React DevTools that emits these events so that its core code does not rely on traversing the tree. This way it would become compatible with incremental reconciler.

File: src/isomorphic/ReactDebugTool.js (hunk 4 above)
Comment (spicyj) [inline src/isomorphic/ReactDebugTool.js]: Yeah.

**Comment by facebook-github-bot-5** [comment]
@spicyj updated the pull request.

**Comment by spicyj** [comment]
Now tracking element, and removed extra debugging code (moved to #6768).

**Comment by gaearon** [comment]
LGTM

**Comment by troydemonbreun** [comment]
Ugh, I guess I totally missed the boat on #6398, I think @gaearon tried to warn me about `ReactComponentTreeDevtool` landing around that time. :-)

**Comment by spicyj** [comment]
@troydemonbreun No worries, your work was still helpful! Thanks for doing most of the work.

**Comment by troydemonbreun** [comment]
@spicyj You are very generous.
